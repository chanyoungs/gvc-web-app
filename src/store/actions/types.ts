import { ExtendedFirestoreInstance, getFirebase } from "react-redux-firebase"
import { ThunkAction } from "redux-thunk"
import { Font } from "src/components/Level1/Dialogs/FontDialog"
import { IBibleRef } from "src/components/Pages/BiblePage"

import { IFBError, INotice, IReport, IReports, TPrayerQueries } from "../../types"
import { AppState } from "../reducers/rootReducer"

// Setup
export type ThunkActionCustom<ReturnType> = ThunkAction<
  ReturnType,
  AppState,
  {
    getFirebase: typeof getFirebase
    // getFirebase: (() => typeof firebase) & typeof getFirebase
    getFirestore: () => ExtendedFirestoreInstance
  },
  AppActions
>

// ServiceWorker Actions
export const SERVICE_WORKER_INIT = "SERVICE_WORKER_INIT"
export interface ServiceWorkerInit {
  type: typeof SERVICE_WORKER_INIT
  payload: boolean
}

export const SERVICE_WORKER_UPDATE = "SERVICE_WORKER_UPDATE"
export interface ServiceWorkerUpdate {
  type: typeof SERVICE_WORKER_UPDATE
  payload: ServiceWorkerRegistration
}

export type ServiceWorkerTypes = ServiceWorkerInit | ServiceWorkerUpdate

// Report actions
export const UPDATE_REPORT = "UPDATE_REPORT"
export interface UpdateReport {
  type: typeof UPDATE_REPORT
  payload: IReport
}

export const UPDATE_BATCH_REPORTS = "UPDATE_BATCH_REPORTS"
export interface UpdateBatchReports {
  type: typeof UPDATE_BATCH_REPORTS
  payload: IReports
}

export const UPDATE_BATCH_REPORTS_ERROR = "UPDATE_BATCH_REPORTS_ERROR"
export interface UpdateBatchReportsError {
  type: typeof UPDATE_BATCH_REPORTS_ERROR
  payload: IFBError
}

export const UPDATE_REPORT_ERROR = "UPDATE_REPORT_ERROR"
export interface UpdateReportError {
  type: typeof UPDATE_REPORT_ERROR
  payload: IFBError
}

export const UPDATE_PRAYER = "UPDATE_PRAYER"
export interface UpdatePrayer {
  type: typeof UPDATE_PRAYER
  payload: { reportId: string; prayer: string }
}

export const UPDATE_PRAYER_ERROR = "UPDATE_PRAYER_ERROR"
export interface UpdatePrayerError {
  type: typeof UPDATE_PRAYER_ERROR
  payload: IFBError
}

export const UPDATE_ATTENDANCE = "UPDATE_ATTENDANCE"
export interface UpdateAttendance {
  type: typeof UPDATE_ATTENDANCE
  payload: { reportId: string; attendance: IReport["attendance"] }
}

export const UPDATE_ATTENDANCE_ERROR = "UPDATE_ATTENDANCE_ERROR"
export interface UpdateAttendanceError {
  type: typeof UPDATE_ATTENDANCE_ERROR
  payload: IFBError
}

export type ReportActionTypes =
  | UpdateReport
  | UpdateBatchReports
  | UpdateBatchReportsError
  | UpdateReportError
  | UpdatePrayer
  | UpdatePrayerError
  | UpdateAttendance
  | UpdateAttendanceError

// Alert actions
export const ALERT_SAVED = "ALERT_SAVED"
export interface AlertSaved {
  type: typeof ALERT_SAVED
  payload: boolean
}

export const ALERT_SAVED_ERROR = "ALERT_SAVED_ERROR"
export interface AlertSavedError {
  type: typeof ALERT_SAVED_ERROR
  payload: IFBError
}

export type AlertActionTypes = AlertSaved | AlertSavedError

// Theme actions
export const UPLOAD_THEME = "UPLOAD_THEME"
export interface UploadTheme {
  type: typeof UPLOAD_THEME
}

export const UPLOAD_THEME_ERROR = "UPLOAD_THEME_ERROR"
export interface UploadThemeError {
  type: typeof UPLOAD_THEME_ERROR
  payload: IFBError
}

export const DELETE_THEME = "DELETE_THEME"
export interface DeleteTheme {
  type: typeof DELETE_THEME
}

export const DELETE_THEME_ERROR = "DELETE_THEME_ERROR"
export interface DeleteThemeError {
  type: typeof DELETE_THEME_ERROR
  payload: IFBError
}

export const SET_CURRENT_THEME_NAME = "SET_CURRENT_THEME_NAME"
export interface SetCurrentThemeName {
  type: typeof SET_CURRENT_THEME_NAME
  payload: string
}

export const SET_CURRENT_THEME_NAME_ERROR = "SET_CURRENT_THEME_NAME_ERROR"
export interface SetCurrentThemeNameError {
  type: typeof SET_CURRENT_THEME_NAME_ERROR
  payload: IFBError
}

export type ThemeActionTypes =
  | UploadTheme
  | UploadThemeError
  | SetCurrentThemeName
  | SetCurrentThemeNameError
  | DeleteTheme
  | DeleteThemeError

// Theme actions
export const UPLOAD_FONT = "UPLOAD_FONT"
export interface UploadFont {
  type: typeof UPLOAD_FONT
}

export const UPLOAD_FONT_ERROR = "UPLOAD_FONT_ERROR"
export interface UploadFontError {
  type: typeof UPLOAD_FONT_ERROR
  payload: IFBError
}

export const DELETE_FONT = "DELETE_FONT"
export interface DeleteFont {
  type: typeof DELETE_FONT
}

export const DELETE_FONT_ERROR = "DELETE_FONT_ERROR"
export interface DeleteFontError {
  type: typeof DELETE_FONT_ERROR
  payload: IFBError
}

export type FontActionTypes =
  | UploadFont
  | UploadFontError
  | DeleteFont
  | DeleteFontError

// Notice actions
export const CREATE_NOTICE = "CREATE_NOTICE"
export interface CreateNotice {
  type: typeof CREATE_NOTICE
  payload: INotice
}

export const CREATE_NOTICE_ERROR = "CREATE_NOTICE_ERROR"
export interface CreateNoticeError {
  type: typeof CREATE_NOTICE_ERROR
  payload: Error
}

export const DELETE_NOTICE = "DELETE_NOTICE"
export interface DeleteNotice {
  type: typeof DELETE_NOTICE
  payload: string
}

export const DELETE_NOTICE_ERROR = "DELETE_NOTICE_ERROR"
export interface DeleteNoticeError {
  type: typeof DELETE_NOTICE_ERROR
  payload: IFBError
}

export type NoticeActionTypes =
  | CreateNotice
  | CreateNoticeError
  | DeleteNotice
  | DeleteNoticeError

// Prayer actions
export const QUERY_PRAYER = "QUERY_PRAYER"
export interface QueryPrayer {
  type: typeof QUERY_PRAYER
  payload: TPrayerQueries
}

export const QUERY_PRAYER_ERROR = "QUERY_PRAYER_ERROR"
export interface QueryPrayerError {
  type: typeof QUERY_PRAYER_ERROR
  payload: IFBError
}

export type PrayerActionTypes = QueryPrayer | QueryPrayerError

// Auth actions
export const SIGN_UP = "SIGN_UP"
export interface SignUp {
  type: typeof SIGN_UP
}

export const SIGN_UP_ERROR = "SIGN_UP_ERROR"
export interface SignUpError {
  type: typeof SIGN_UP_ERROR
  payload: IFBError
}

export const MEMBER_PROFILE_CREATED = "MEMBER_PROFILE_CREATED"
export interface MemberProfileCreated {
  type: typeof MEMBER_PROFILE_CREATED
}

export const MEMBER_PROFILE_CREATED_ERROR = "MEMBER_PROFILE_CREATED_ERROR"
export interface MemberProfileCreatedError {
  type: typeof MEMBER_PROFILE_CREATED_ERROR
  payload: IFBError
}

export const SIGN_IN = "SIGN_IN"
export interface SignIn {
  type: typeof SIGN_IN
}

export const SIGN_IN_ERROR = "SIGN_IN_ERROR"
export interface SignInError {
  type: typeof SIGN_IN_ERROR
  payload: IFBError
}

export const REMEMBER_ME = "REMEMBER_ME"
export interface RememberMe {
  type: typeof REMEMBER_ME
  payload: boolean
}

export const REMEMBER_ME_ERROR = "REMEMBER_ME_ERROR"
export interface RememberMeError {
  type: typeof REMEMBER_ME_ERROR
  payload: IFBError
}

export const RESET_PASSWORD = "RESET_PASSWORD"
export interface ResetPassword {
  type: typeof RESET_PASSWORD
}

export const RESET_PASSWORD_ERROR = "RESET_PASSWORD_ERROR"
export interface ResetPasswordError {
  type: typeof RESET_PASSWORD_ERROR
  payload: IFBError
}

export const SIGN_OUT = "SIGN_OUT"
export interface SignOut {
  type: typeof SIGN_OUT
}

export const SIGN_OUT_ERROR = "SIGN_OUT_ERROR"
export interface SignOutError {
  type: typeof SIGN_OUT_ERROR
  payload: IFBError
}

export const EDIT_PROFILE = "EDIT_PROFILE"
export interface EditProfile {
  type: typeof EDIT_PROFILE
}

export const EDIT_PROFILE_ERROR = "EDIT_PROFILE_ERROR"
export interface EditProfileError {
  type: typeof EDIT_PROFILE_ERROR
  payload: IFBError
}

export const UPLOAD_PHOTO = "UPLOAD_PHOTO"
export interface UploadPhoto {
  type: typeof UPLOAD_PHOTO
}

export const UPLOAD_PHOTO_ERROR = "UPLOAD_PHOTO_ERROR"
export interface UploadPhotoError {
  type: typeof UPLOAD_PHOTO_ERROR
  payload: Error
}

export const DELETE_PHOTO = "DELETE_PHOTO"
export interface DeletePhoto {
  type: typeof DELETE_PHOTO
}

export const DELETE_PHOTO_ERROR = "DELETE_PHOTO_ERROR"
export interface DeletePhotoError {
  type: typeof DELETE_PHOTO_ERROR
  payload: IFBError
}

export type AuthActionTypes =
  | SignUp
  | SignUpError
  | MemberProfileCreated
  | MemberProfileCreatedError
  | SignIn
  | SignInError
  | RememberMe
  | RememberMeError
  | ResetPassword
  | ResetPasswordError
  | SignOut
  | SignOutError
  | EditProfile
  | EditProfileError
  | UploadPhoto
  | UploadPhotoError
  | DeletePhoto
  | DeletePhotoError
  | DeleteNotice
  | DeleteNoticeError
  | UploadTheme
  | UploadThemeError
  | SetCurrentThemeName
  | SetCurrentThemeNameError
  | DeleteTheme
  | DeleteThemeError

// AppBar Actions
export const SEARCH_ON_CHANGE = "SEARCH_ON_CHANGE"
export interface SearchOnChange {
  type: typeof SEARCH_ON_CHANGE
  payload: string
}

export const SET_DRAWER_OPEN = "SET_DRAWER_OPEN"
export interface SetDrawerOpen {
  type: typeof SET_DRAWER_OPEN
  payload: boolean
}

export type AppBarActionTypes = SearchOnChange | SetDrawerOpen

// Bible Actions
export const SET_BIBLE_REFERENCE = "SET_BIBLE_REFERENCE"
export interface SetBibleReference {
  type: typeof SET_BIBLE_REFERENCE
  payload: IBibleRef
}

export const SET_BIBLE_REFERENCE_ERROR = "SET_BIBLE_REFERENCE_ERROR"
export interface SetBibleReferenceError {
  type: typeof SET_BIBLE_REFERENCE_ERROR
  payload: IFBError
}

export type BibleActionTypes = SetBibleReference | SetBibleReferenceError

// All Action types
export type AppActions =
  | NoticeActionTypes
  | PrayerActionTypes
  | AuthActionTypes
  | AppBarActionTypes
  | BibleActionTypes
  | ThemeActionTypes
  | FontActionTypes
  | AlertActionTypes
  | ReportActionTypes
  | ServiceWorkerTypes
