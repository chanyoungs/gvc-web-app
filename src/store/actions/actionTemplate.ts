// "cta" is a snippet prefix for the template below
import { ThunkActionCustom } from "../../types/actions"

export const createNotice = (): ThunkActionCustom<void> => (
  dispatch,
  getState,
  { getFirestore, getFirebase }
) => {
  const firestore = getFirestore()
  const firebase = getFirebase()

  // firestore
  //   .collection("notices")
  //   .add({ ...notice, createdAt: new Date() })
  //   .then(() => {
  //     dispatch({ type: "CREATE_NOTICE", payload: notice })
  //   })
  //   .catch((error: Error) => {
  //     dispatch({ type: "CREATE_NOTICE_ERROR", payload: error })
  //   })
}
