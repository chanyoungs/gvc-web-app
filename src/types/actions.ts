import { ExtendedFirestoreInstance, getFirebase } from "react-redux-firebase"
import { ThunkAction } from "redux-thunk"

import { IFBError, INotice } from "."
import { AppState } from "../store/reducers/rootReducer"

// Setup
export type ThunkActionCustom<ReturnType> = ThunkAction<
  ReturnType,
  AppState,
  {
    getFirebase: typeof getFirebase
    // getFirebase: (() => typeof firebase) & typeof getFirebase
    getFirestore: () => ExtendedFirestoreInstance
  },
  AppActions
>

// Notice actions
export const CREATE_NOTICE = "CREATE_NOTICE"
export interface CreateNotice {
  type: typeof CREATE_NOTICE
  payload: INotice
}

export const CREATE_NOTICE_ERROR = "CREATE_NOTICE_ERROR"
export interface CreateNoticeError {
  type: typeof CREATE_NOTICE_ERROR
  payload: Error
}

export type NoticeActionTypes = CreateNotice | CreateNoticeError

// Auth actions
export const SIGN_UP = "SIGN_UP"
export interface SignUp {
  type: typeof SIGN_UP
}

export const SIGN_UP_ERROR = "SIGN_UP_ERROR"
export interface SignUpError {
  type: typeof SIGN_UP_ERROR
  payload: IFBError
}

export const MEMBER_PROFILE_CREATED = "MEMBER_PROFILE_CREATED"
export interface MemberProfileCreated {
  type: typeof MEMBER_PROFILE_CREATED
}

export const MEMBER_PROFILE_CREATED_ERROR = "MEMBER_PROFILE_CREATED_ERROR"
export interface MemberProfileCreatedError {
  type: typeof MEMBER_PROFILE_CREATED_ERROR
  payload: IFBError
}

export const SIGN_IN = "SIGN_IN"
export interface SignIn {
  type: typeof SIGN_IN
}

export const SIGN_IN_ERROR = "SIGN_IN_ERROR"
export interface SignInError {
  type: typeof SIGN_IN_ERROR
  payload: IFBError
}

export const REMEMBER_ME = "REMEMBER_ME"
export interface RememberMe {
  type: typeof REMEMBER_ME
  payload: boolean
}

export const REMEMBER_ME_ERROR = "REMEMBER_ME_ERROR"
export interface RememberMeError {
  type: typeof REMEMBER_ME_ERROR
  payload: IFBError
}

export const SIGN_OUT = "SIGN_OUT"
export interface SignOut {
  type: typeof SIGN_OUT
}

export const SIGN_OUT_ERROR = "SIGN_OUT_ERROR"
export interface SignOutError {
  type: typeof SIGN_OUT_ERROR
  payload: IFBError
}

export type AuthActionTypes =
  | SignUp
  | SignUpError
  | MemberProfileCreated
  | MemberProfileCreatedError
  | SignIn
  | SignInError
  | RememberMe
  | RememberMeError
  | SignOut
  | SignOutError

// All Action types
export type AppActions = NoticeActionTypes | AuthActionTypes
