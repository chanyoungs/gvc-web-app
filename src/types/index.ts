// ---Notices---
export interface INotice {
  title: string
  content: string
}

export interface INoticeWithMeta extends INotice {
  id: string
  createdAt: Date
}

// ---Auth---
// Elements
export interface IEmail {
  email: string
}
export interface ISetSubmitting {
  setSubmitting: (isSubmitting: boolean) => void
}
export interface IOpenAlert {
  openAlert: () => void
}
export interface ICredentials extends IEmail {
  password: string
}

// Auth Interface
export interface IResetPassword extends IEmail, ISetSubmitting, IOpenAlert {}
export interface ISignIn extends ICredentials, ISetSubmitting {
  rememberMe: boolean
  redirectOnSignIn: () => void
}
export interface ISignUp extends ICredentials, ISetSubmitting, IOpenAlert {
  name: string
  dob: Date | null
  agreeTAndC: boolean
}

// Firebase Error Interface
export interface IFBError {
  code: string
  message: string
}

// ---Pages---
export type Paths =
  | "/"
  | "/myaccount"
  | "/prayers"
  | "/members"
  | "/calendar"
  | "/public"
  | "/private"
