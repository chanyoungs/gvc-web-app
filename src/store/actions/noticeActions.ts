import { ThunkActionCustom } from '../../types/actions'
import { INotice, INoticeWithMeta } from './../../types'

export const createNotice = (notice: INotice): ThunkActionCustom<void> => (
  dispatch,
  getState,
  { getFirestore, getFirebase }
) => {
  const firestore = getFirestore()
  const firebase = getFirebase()

  firestore
    .collection('notices')
    .add({ ...notice, createdAt: new Date() })
    .then(() => {
      dispatch({ type: 'CREATE_NOTICE', payload: notice })
    })
    .catch((error: Error) => {
      dispatch({ type: 'CREATE_NOTICE_ERROR', payload: error })
    })
}

export const deleteNotice = (id: string): ThunkActionCustom<void> => (
  dispatch,
  getState,
  { getFirestore, getFirebase }
) => {
  const firestore = getFirestore()

  console.log('deleteNotice', id)
  let deleteDoc = firestore
    .collection('notices')
    .doc(id)
    .delete()
    .then(() => {
      dispatch({ type: 'DELETE_NOTICE', payload: id })
    })
    .catch((error: Error) => {
      dispatch({ type: 'CREATE_NOTICE_ERROR', payload: error })
    })

  console.log('deleteDoc', deleteDoc)
}

export const updateNotice = (
  notice: INoticeWithMeta
): ThunkActionCustom<void> => (
  dispatch,
  getState,
  { getFirestore, getFirebase }
) => {
  const firestore = getFirestore()
  const firebase = getFirebase()

  firestore
    .collection('notices')
    .doc(notice.id)
    .set(notice)
    .then(() => {
      dispatch({ type: 'CREATE_NOTICE', payload: notice })
    })
    .catch((error: Error) => {
      dispatch({ type: 'CREATE_NOTICE_ERROR', payload: error })
    })
}
