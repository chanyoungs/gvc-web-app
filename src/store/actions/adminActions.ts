import { ICell, ICells, IMemberDownload, IMemberUpload } from "src/types"
import { v4 as uuid } from "uuid"

import { ThunkActionCustom } from "./types"

export interface UpdateMemberCellProps {
  memberId: string
  newCellId: string
  callback?: () => void
}

export const updateMemberCell =
  ({
    memberId,
    newCellId,
    callback,
  }: UpdateMemberCellProps): ThunkActionCustom<void> =>
  async (dispatch, getState, { getFirestore, getFirebase }) => {
    const firestore = getFirestore()

    try {
      await firestore
        .collection("members")
        .doc(memberId)
        .update({ cell: newCellId, cellRequest: newCellId })
      console.log("Cell (re)allocation success!")
      callback && callback()
    } catch (error) {
      console.error("Cell (re)allocation fail!", error)
    }
  }

export interface AddNewCellProps {
  name: string
  callback?: () => void
}

export const addNewCell =
  ({ name, callback }: AddNewCellProps): ThunkActionCustom<void> =>
  async (dispatch, getState, { getFirestore, getFirebase }) => {
    const firestore = getFirestore()
    const newCellId = uuid()
    const newCell: ICell = {
      id: newCellId,
      leaders: [],
      name,
    }
    try {
      await firestore
        .collection("cells")
        .doc("cells")
        .update({
          [newCellId]: newCell,
        })
      console.log(`Cell ${name} added successfully!`)
      callback && callback()
    } catch (error) {
      console.error("Cell addition fail!", error)
    }
  }

// export const updateMemberCell =
// ({
//   memberId,
//   currentCellId,
//   newCellId,
//   callback,
// }: UpdateMemberCellProps): ThunkActionCustom<void> =>
// async (dispatch, getState, { getFirestore, getFirebase }) => {
//   const firestore = getFirestore()
//   const cellsRef = firestore.collection("cells").doc("cells")

//   try {
//     await firestore.runTransaction(async (transaction) => {
//       const cellsSnapshot = await transaction.get(cellsRef)
//       const cells: ICells | undefined = cellsSnapshot.data()

//       if (cells) {
//         // If current cell exists, remove member from previous cell
//         if (currentCellId in cells) {
//           transaction.update(cellsRef, {
//             [`${currentCellId}.members`]:
//               firestore.FieldValue.arrayRemove(memberId),
//           })
//         }

//         // If member doesn't exist in the new cell already, add member to the new cell
//         if (!cells[newCellId].members.includes(currentCellId)) {
//           // Add member to new cell
//           transaction.update(cellsRef, {
//             [`${newCellId}.members`]:
//               firestore.FieldValue.arrayUnion(memberId),
//           })
//         }

//         // Update member profile's cell
//         transaction.update(firestore.collection("members").doc(memberId), {
//           cell: newCellId,
//         })
//       }
//     })
//     console.log("Cell (re)allocation success!")
//     callback && callback()
//   } catch (error) {
//     console.error("Cell (re)allocation fail!", error)
//   }
// }
