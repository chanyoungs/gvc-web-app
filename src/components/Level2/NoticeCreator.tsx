import Button from '@material-ui/core/Button'
import Textfield from '@material-ui/core/TextField'
import React, { Fragment } from 'react'
import { useDispatch } from 'react-redux'

import { createNotice } from './../../store/actions/noticeActions'
import { INotice } from './../../types'

export const NoticeCreator: React.FC = () => {
  const [values, setValues] = React.useState<INotice>({
    title: '',
    content: ''
  })

  const onChange = (name: keyof INotice) => (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    setValues({ ...values, [name]: event.target.value })
  }

  const dispatch = useDispatch()
  const onSubmit = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    e.preventDefault()
    setValues({
      title: '',
      content: ''
    })

    dispatch(createNotice(values))
  }

  return (
    <Fragment>
      <Textfield
        data-testid="title"
        label="Title"
        fullWidth
        onChange={onChange('title')}
        value={values.title}
      />
      <Textfield
        data-testid="content"
        id="standard-multiline-flexible"
        fullWidth
        multiline
        rowsMax="4"
        label="Content"
        onChange={onChange('content')}
        value={values.content}
      />
      <Button data-testid="submit" onClick={onSubmit}>
        Create new notice
      </Button>
    </Fragment>
  )
}
