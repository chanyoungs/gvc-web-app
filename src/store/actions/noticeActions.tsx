import { Dispatch } from "redux"

import { State as INotice } from "./../../components/Level2/NoticeCreator"

export const createNotice = (notice: INotice) => {
  return (dispatch: Dispatch, getState: any, { getFirestore }: any) => {
    const firestore = getFirestore()
    firestore
      .collection("notices")
      .add({ ...notice, createdAt: new Date() })
      .then(() => {
        dispatch({ type: "CREATE_NOTICE", payload: notice })
      })
      .catch((err: any) => {
        dispatch({ type: "CREATE_NOTICE_ERROR", payload: err })
      })
  }
}
