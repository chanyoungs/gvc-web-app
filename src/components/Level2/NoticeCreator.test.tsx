import React from 'react'
import Enzyme, { mount } from 'enzyme'
import { NoticeCreator } from './NoticeCreator'
import EnzymeAdapter from 'enzyme-adapter-react-16'
import { useDispatch } from 'react-redux'
import { ThunkActionCustom } from '../../types/actions'
import { createNotice } from './../../store/actions/noticeActions'

Enzyme.configure({
  adapter: new EnzymeAdapter(),
  disableLifecycleMethods: true
})

const mockDispatch = jest.fn()
jest.mock('react-redux', () => ({
  useDispatch: jest.fn().mockReturnValue(mockDispatch)
}))

jest.mock('../../types/actions', () => ({
  createNotice: jest.fn(x => x)
}))

// const mockUseDispatch = useDispatch as jest.Mock
// const mockCreateNotice = createNotice as jest.Mock
// const mockDispatch = jest.fn(() =>
//   mockCreateNotice({ title: 'new title', content: 'new content' })
// )

describe('<NoticeCreator />', () => {
  // mockUseDispatch.mockReturnValue(mockDispatch)

  const container = mount(<NoticeCreator />)

  it('should match the snapshot', () => {
    expect(container.html()).toMatchSnapshot()
  })

  it('should have an title field', () => {
    expect(
      container.find('[data-testid="title"] input[type="text"]').length
    ).toEqual(1)
  })

  it('should have an content field', () => {
    expect(
      container.find('[data-testid="content"] textarea').length
    ).toBeGreaterThan(1)
  })

  it('should set the title value on change event', () => {
    container
      .find('[data-testid="title"] input[type="text"]')
      .simulate('change', {
        target: {
          value: 'some new title'
        }
      })

    expect(
      container.find('[data-testid="title"] input[type="text"]').prop('value')
    ).toBe('some new title')
  })

  it('should set the content value on change event', () => {
    container
      .find('[data-testid="content"] textarea')
      .at(0)
      .simulate('change', {
        target: {
          value: 'some new content'
        }
      })

    expect(
      container
        .find('[data-testid="content"] textarea')
        .at(0)
        .prop('value')
    ).toBe('some new content')
  })

  it('should call the add notice function on submit button click', () => {
    container
      .find('[data-testid="title"] input[type="text"]')
      .simulate('change', {
        target: {
          value: 'some new title'
        }
      })
    container
      .find('[data-testid="content"] textarea')
      .at(0)
      .simulate('change', {
        target: {
          value: 'some new content'
        }
      })

    container.find('button[type="button"]').simulate('click')

    expect(mockDispatch).toHaveBeenCalledTimes(1)
    //dispatch 제대로 호출됐는지 체크 with argument
    expect(mockDispatch).toHaveBeenCalledWith(
      createNotice({ title: 'new title', content: 'new content' })
    )

    expect(
      container
        .find('[data-testid="content"] textarea')
        .at(0)
        .prop('value')
    ).toBe('')

    expect(
      container.find('[data-testid="title"] input[type="text"]').prop('value')
    ).toBe('')
  })
})
