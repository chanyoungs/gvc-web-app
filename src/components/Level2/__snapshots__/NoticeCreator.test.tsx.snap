// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoticeCreator /> should match the snapshot 1`] = `"<div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth\\" data-testid=\\"title\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated\\" data-shrink=\\"false\\">Title</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl\\"><input aria-invalid=\\"false\\" type=\\"text\\" class=\\"MuiInputBase-input MuiInput-input\\" value=\\"\\"></div></div><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth\\" data-testid=\\"content\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated\\" data-shrink=\\"false\\" for=\\"standard-multiline-flexible\\" id=\\"standard-multiline-flexible-label\\">Content</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-multiline MuiInput-multiline\\"><textarea rows=\\"1\\" aria-invalid=\\"false\\" id=\\"standard-multiline-flexible\\" class=\\"MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline\\" style=\\"height: 0px; overflow: hidden;\\"></textarea><textarea aria-hidden=\\"true\\" class=\\"MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline\\" readonly=\\"\\" tabindex=\\"-1\\" style=\\"visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 100%;\\"></textarea></div></div><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text\\" tabindex=\\"0\\" type=\\"button\\" data-testid=\\"submit\\"><span class=\\"MuiButton-label\\">Create new notice</span><span class=\\"MuiTouchRipple-root\\"></span></button>"`;
